### PR TITLE
Use CanonicalSerialization of RawTransaction between client and validator

### PR DESCRIPTION
**Pull request by davidiw**
## Motivation

RawTransactions are signed by clients and transmitted to all validators. Validators verify their authenticity by comparing the bundled signature to the RawTransaction. RawTransaction uses protobuf requiring that the validator and client keep both a copy of the original client serialized bytes and the deserialized object, due to RawTransaction's use of protobuf.

This diff eliminates the reliance on protobuf but at the same time introduce breaking changes for current clients and validators.

## Test Plan

cargo test in libra/
verified that all code paths in core libra workflow adopt the changes correctly in the various integration and unit tests included there in

## Related PRs

https://github.com/libra/libra/pull/724

### CODE REVIEW

File: libra_swarm/src/swarm.rs [hunk 1]
```diff
@@ -273,7 +273,7 @@ impl LibraSwarm {
                 Ok(swarm) => {
                     return swarm;
                 }
-                Err(e) => error!("Error launching swarm: {}", e),
+                Err(e) => panic!("Error launching swarm: {}", e),
```
Comment (zamsden) [inline libra_swarm/src/swarm.rs]: This seems unrelated.

Comment (davidiw) [reply libra_swarm/src/swarm.rs]: Good call, my machine had some transient issue I was trying to debug related to writes to disk. Cleaning up now...

**Comment by kphfb** [comment]
It looks like you have some conflicts, can you take a look at those?

**Comment by davidiw** [comment]
> It looks like you have some conflicts, can you take a look at those?

Hmm... it is disappointing that github doesn't leverage the commit from master that the diff was based upon...

Fixed!

File: common/canonical_serialization/README.md [hunk 2]
```diff
@@ -0,0 +1,471 @@
...
+
+## Background
+
+In Libra, participants pass around messages or data structures that often times need to be signed by one party and verified by many. Serialization is the process of converting a message into a byte array. Most serialization approaches support loose standards such that two implementations can produce two different byte streams that would represent the identical message. As a result, messages that are passed between participants must either retain the original serialized bytes or risk losing the ability to verify the message. This creates a burden programmatically by requiring the code to maintain a copy of the bytes after deserializing the object often leads to confusion about the safety and correctness of this approach. While there exist a handful of existing deterministic serialization formats, e.g. XDR, there exists no mainstream, well maintained, and available deterministic serialization. To address this, we propose Libra Canonical Serialization that defines a deterministic means for translating a message into bytes.
```
[the remark this answers is not in the document]
Comment (davidiw) [reply common/canonical_serialization/README.md]: ASN.1!!! Good point! Will fix in revised copy.

File: common/canonical_serialization/README.md [hunk 3]
```diff
@@ -0,0 +1,471 @@
...
+}
+```
+
+### Examples
+
+**AccountAddress**
+
+String representation:
+```
+ca820bf9305eb97d0d784f71b3955457fbf6911f5300ceaa5d7e8621529eae19
```
[the remark this answers is not in the document]
Comment (davidiw) [reply common/canonical_serialization/README.md]: Agreed, the AccountAddress though is a fixed 32-byte structure, thus complicating simplification for this specific type.

File: common/canonical_serialization/README.md [hunk 4]
```diff
@@ -0,0 +1,471 @@
...
+```
+enum Option {
+  option0(u32),
+  option1(u64)
+}
+```
+
+wherein the enum object can only representation one of these options. This is similar to a union in C.
+
+In LCS, each option is mapped to a specific unsigned 32-bit integer followed by optionally serialized data if the type predicates it.
```
[the remark this answers is not in the document]
Comment (davidiw) [reply common/canonical_serialization/README.md]: asserts

though I'll use more sensible diction: "if the type has an associated value"

thanks for calling this out.

**Comment by huitseeker** [comment]
This PR is getting relatively unwieldy, so that reviewing #622 is becoming hard for size reasons. Can we focus on this one first, rebasing it an merging it before doing further work ?

**Comment by davidiw** [comment]
> This PR is getting relatively unwieldy, so that reviewing #622 is becoming hard for size reasons. Can we focus on this one first, rebasing it an merging it before doing further work ?

I think the goal is to get #622 completed first as when we were working on this effort we noticed that there were many deficiencies in place that need to be resolved before we take on rolling out any breaking changes.

With respect to #622 or LCS in general, the plan is:
- Refactor LCS to modernize it (taking on a lot of your suggestions from an early PR and standardizing the API)
- Improving test coverage via proptest (again taking on your suggestion)
- Eliminating access to raw bytes -- all bytes must be serialized with a prefixed length
- Adding a version field prior to all structs

The only feature currently missing is the version field prior to structs. I'm pondering a low cost fashion to do so to support our current structs as they are all effectively version 0.

All those changes should only break the protocol between validators, so I'm happy to do whatever to make reviewers live's easy to expedite reviews (i.e., multiple PRs). Suggestions / feedback is appreciated.

Once those changes are in place, this diff can be rebased and landed sometime Tuesday or later, so we give time for partners to fix their transaction generation logic in their clients.

I wish the UI gave me the ability to specify that this PR is not ready for review.

File: types/src/transaction.rs [hunk 5]
```diff
@@ -401,8 +401,8 @@ impl SignedTransaction {
             raw_txn: raw_txn.clone(),
             public_key,
             signature,
-            // In real world raw_txn should be derived from raw_txn_bytes, not the opposite.
-            raw_txn_bytes: raw_txn.into_proto_bytes().expect("Should convert."),
+            raw_txn_bytes: SimpleSerializer::<Vec<u8>>::serialize(&raw_txn)
```
Comment (kphfb) [inline types/src/transaction.rs]: Annoyingly, I can't seem to comment on unchanged lines, so I'll add the comment here.  Why do we still have raw_txn_bytes in SignedTransaction?  It seems like we would only really need it in case we wanted to be able to provide the client with an error message like "Incorrectly serialized"

Comment (davidiw) [reply types/src/transaction.rs]: Agreed. This will require non-trivial fixes to the core of Libra. I didn't want to incur the wrath of making yet another massive PR. So I'll hopefully have that ready Monday afternoon or Tuesday. So we can land it shortly after the next cut.

Comment (davidiw) [reply types/src/transaction.rs]: Actually... maybe we can break it up into two things... one is to fix the proto / grpc call. The second would be to eliminate this field on the rust data structure -- that way we can easily make the internal change without impact clients.
